src/EventStore.Core/Services/NodeStateListenerService.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Text.Json;
+using EventStore.Core.Bus;
+using EventStore.Core.Data;
+using EventStore.Core.Messages;
+using EventStore.Core.Services.Storage;
+using EventStore.Core.TransactionLog.LogRecords;
+
+namespace EventStore.Core.Services;
+
+// threading: we expect to handle one StateChangeMessage at a time, but Reads can happen concurrently
+// with those handlings and with other reads.
+public class NodeStateListenerService :
+	IInMemoryStreamReader,
+	IHandle<SystemMessage.StateChangeMessage> {
+	private readonly IPublisher _publisher;
+	public const string EventType = "NodeStateChanged";
```

Comment:
Should this be an admin event type?
```suggestion
	public const string EventType = "$NodeStateChanged";
```